ssh/controlmaster.go

```diff
@@ -0,0 +1,233 @@
+type controlMasterAliveMessage struct {
+	PacketType uint32 // MUX_S_ALIVE
+	RequestID  uint32
+	ServerPID  uint32
+}
+
+// controlMasterHelloMessage is used to ask/send the protocol version
+type controlMasterHelloMessage struct {
+	PacketType      uint32 // MUX_MSG_HELLO
+	ProtocolVersion uint32
+}
+
+// DialControlMaster starts a client connection to the given SSH server.
+// It is a convenience function that connects to the given unix socket,
+// initiates the ControlMaster proxy prottocol, and then sets up a Client.
+// For access to incoming channels and requests, use net.Dial with
+// NewControlMasterClientConn instead.
+func DialControlMaster(path string) (*Client, error) {
+	conn, err := net.Dial("unix", path)
+	if err != nil {
+		return nil, err
+	}
+	c, chans, reqs, err := NewControlMasterClientConn(conn)
+	if err != nil {
+		return nil, err
+	}
+	return NewClient(c, chans, reqs), nil
+}
+
+// NewControlMasterClientConn establishes a ControlMaster connection
+// using c as the underlying transport. The Request and NewChannel
+// channels must be serviced or the connection will hang.
+func NewControlMasterClientConn(c net.Conn) (Conn, <-chan NewChannel, <-chan *Request, error) {
+	// We never populate transport, just mux since there is no handshake needed
+	conn := &connection{
+		sshConn: sshConn{conn: c},
+	}
+
+	// Instead of calling clientHandshake, call our own setup function
+	if err := setupControlMaster(&controlMasterConnection{
+		rw: c,
+	}); err != nil {
+		return nil, nil, nil, err
+	}
+
+	// Instead of using transport we use our own impl packetConn instead
+	conn.mux = newMux(&controlMasterProxyConnection{
+		rwc: c,
+	})
+	return conn, conn.mux.incomingChannels, conn.mux.incomingRequests, nil
+}
+
+// controlMasterConnection implements the control master protocol
+type controlMasterConnection struct {
+	rw io.ReadWriter
+}
+
+// writePacket prefixes the length to the packet and writes it
+func (c *controlMasterConnection) writePacket(packet interface{}) error {
+	// Marshal the packet as standard SSH packets
+	msg := Marshal(packet)
+	// We need to calculate the length field
+	msgLenBytes := make([]byte, 4)
+	binary.BigEndian.PutUint32(msgLenBytes, uint32(len(msg)))
+	msg = append(msgLenBytes, msg...)
+	n, err := c.rw.Write(msg)
+	if err != nil {
+		return err
+	}
+	if n != len(msg) {
+		return errors.New("Unable to write all bytes in message")
+	}
+	return nil
+}
+
+// readPacket reads a packet based on an initial length value
+func (c *controlMasterConnection) readPacket(packet interface{}) error {
+	// Read the response length
+	respLenBuf := make([]byte, 4)
+	n1, err := c.rw.Read(respLenBuf)
+	if err != nil {
+		return err
+	}
+	if n1 != 4 {
+		return errors.New("Unable to read entire response packet length")
+	}
+	respLen := binary.BigEndian.Uint32(respLenBuf)
+	// Then read the response
+	resp := make([]byte, respLen)
+	n2, err := c.rw.Read(resp)
```

Comment:
Use `io.ReadFull` so that you'll get a full buffer even if it requires calling Read more than once.